File: mobile_apps/ios_app/TigerWallet/Services/NFTMarketplaceService.swift
//
//  NFTMarketplaceService.swift
//  TigerWallet
//
//  Production-ready NFT marketplace functionality
//  Buy, sell, create listings, and trade NFTs
//

import Foundation

// MARK: - NFT Models

struct NFTCollection: Codable, Identifiable {
    let id: String
    let address: String
    let name: String
    let symbol: String
    let description: String?
    let imageUrl: String?
    let bannerUrl: String?
    let floorPrice: Double
    let totalSupply: Int
    let owners: Int
    let volume24h: Double
    let chain: String
}

struct NFTListing: Codable, Identifiable {
    let id: String
    let tokenId: String
    let collectionAddress: String
    let seller: String
    let price: Double
    let priceToken: String
    let expirationTime: Int
    let chain: String
    let imageUrl: String?
    let name: String?
}

struct NFTItem: Codable, Identifiable {
    let id: String
    let tokenId: String
    let collectionAddress: String
    let name: String
    let description: String?
    let imageUrl: String?
    let animationUrl: String?
    let attributes: [NFTAttribute]?
    let owner: String?
    let price: Double?
    let chain: String
}

struct NFTAttribute: Codable {
    let traitType: String
    let value: String
}

struct NFTTrade: Codable, Identifiable {
    let id: String
    let tokenId: String
    let collectionAddress: String
    let buyer: String
    let seller: String
    let price: Double
    let timestamp: Int
    let txHash: String
    let chain: String
}

// MARK: - Marketplace Service

class NFTMarketplaceService {
    
    static let shared = NFTMarketplaceService()
    
    private let session = URLSession.shared
    
    // MARK: - Collection Methods
    
    /// Get collection by address
    func getCollection(address: String, chain: String = "ethereum") async throws -> NFTCollection? {
        let url = URL(string: "http://localhost:8443/api/v1/nft/collections/\(address)?chain=\(chain)")!
        
        let (data, response) = try await session.data(from: url)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return nil
        }
        
        let collection = try JSONDecoder().decode(NFTCollection.self, from: data)
        return collection
    }
    
    /// Search collections
    func searchCollections(query: String, chain: String = "ethereum", limit: Int = 20) async throws -> [NFTCollection] {
        var components = URLComponents(string: "http://localhost:8443/api/v1/nft/collections/search")!
        components.queryItems = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "chain", value: chain),
            URLQueryItem(name: "limit", value: String(limit))
        ]
        
        let (data, response) = try await session.data(from: components.url!)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return []
        }
        
        let collections = try JSONDecoder().decode([NFTCollection].self, from: data)
        return collections
    }
    
    // MARK: - NFT Methods
    
    /// Get NFTs for a collection
    func getNFTs(collectionAddress: String, chain: String = "ethereum", limit: Int = 50, offset: Int = 0) async throws -> [NFTItem] {
        var components = URLComponents(string: "http://localhost:8443/api/v1/nft/assets")!
        components.queryItems = [
            URLQueryItem(name: "collection", value: collectionAddress),
            URLQueryItem(name: "chain", value: chain),
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "offset", value: String(offset))
        ]
        
        let (data, response) = try await session.data(from: components.url!)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return []
        }
        
        let nfts = try JSONDecoder().decode([NFTItem].self, from: data)
        return nfts
    }
    
    /// Get NFTs for a wallet
    func getUserNFTs(walletAddress: String, chain: String = "ethereum") async throws -> [NFTItem] {
        let url = URL(string: "http://localhost:8443/api/v1/nft/owners/\(walletAddress)?chain=\(chain)")!
        
        let (data, response) = try await session.data(from: url)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return []
        }
        
        let nfts = try JSONDecoder().decode([NFTItem].self, from: data)
        return nfts
    }
    
    // MARK: - Listing Methods
    
    /// Get listings for a collection
    func getListings(collectionAddress: String, chain: String = "ethereum", limit: Int = 50) async throws -> [NFTListing] {
        var components = URLComponents(string: "http://localhost:8443/api/v1/nft/listings")!
        components.queryItems = [
            URLQueryItem(name: "collection", value: collectionAddress),
            URLQueryItem(name: "chain", value: chain),
            URLQueryItem(name: "limit", value: String(limit))
        ]
        
        let (data, response) = try await session.data(from: components.url!)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return []
        }
        
        let listings = try JSONDecoder().decode([NFTListing].self, from: data)
        return listings
    }
    
    /// Get user's listings
    func getUserListings(walletAddress: String, chain: String = "ethereum") async throws -> [NFTListing] {
        let url = URL(string: "http://localhost:8443/api/v1/nft/listings/\(walletAddress)?chain=\(chain)")!
        
        let (data, response) = try await session.data(from: url)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return []
        }
        
        let listings = try JSONDecoder().decode([NFTListing].self, from: data)
        return listings
    }
    
    // MARK: - Trading Methods
    
    /// Create a listing (sell NFT)
    func createListing(
        walletAddress: String,
        collectionAddress: String,
        tokenId: String,
        price: Double,
        priceToken: String = "ETH",
        chain: String = "ethereum"
    ) async throws -> String {
        let url = URL(string: "http://localhost:8443/api/v1/nft/listings")!
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "wallet_address": walletAddress,
            "collection_address": collectionAddress,
            "token_id": tokenId,
            "price": price,
            "price_token": priceToken,
            "chain": chain
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw NFTError.transactionFailed
        }
        
        let result = try JSONDecoder().decode(ListingResult.self, from: data)
        return result.listingId
    }
    
    /// Cancel a listing
    func cancelListing(walletAddress: String, listingId: String, chain: String = "ethereum") async throws -> Bool {
        let url = URL(string: "http://localhost:8443/api/v1/nft/listings/\(listingId)")!
        
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "wallet_address": walletAddress,
            "chain": chain
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (_, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw NFTError.transactionFailed
        }
        
        return true
    }
    
    /// Buy NFT (fulfill listing)
    func buyNFT(buyerAddress: String, listingId: String, chain: String = "ethereum") async throws -> String {
        let url = URL(string: "http://localhost:8443/api/v1/nft/purchase")!
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "buyer_address": buyerAddress,
            "listing_id": listingId,
            "chain": chain
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw NFTError.transactionFailed
        }
        
        let result = try JSONDecoder().decode(TransactionResult.self, from: data)
        return result.txHash
    }
    
    // MARK: - Offer Methods
    
    /// Make an offer
    func makeOffer(
        makerAddress: String,
        collectionAddress: String,
        tokenId: String,
        price: Double,
        priceToken: String = "ETH",
        expirationTime: Int,
        chain: String = "ethereum"
    ) async throws -> String {
        let url = URL(string: "http://localhost:8443/api/v1/nft/offers")!
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "maker_address": makerAddress,
            "collection_address": collectionAddress,
            "token_id": tokenId,
            "price": price,
            "price_token": priceToken,
            "expiration_time": expirationTime,
            "chain": chain
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw NFTError.transactionFailed
        }
        
        let result = try JSONDecoder().decode(OfferResult.self, from: data)
        return result.offerId
    }
    
    /// Accept an offer
    func acceptOffer(sellerAddress: String, offerId: String, chain: String = "ethereum") async throws -> String {
        let url = URL(string: "http://localhost:8443/api/v1/nft/offers/\(offerId)/accept")!
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "seller_address": sellerAddress,
            "chain": chain
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw NFTError.transactionFailed
        }
        
        let result = try JSONDecoder().decode(TransactionResult.self, from: data)
        return result.txHash
    }
    
    // MARK: - Trade History
    
    /// Get trade history for a collection
    func getCollectionTrades(collectionAddress: String, chain: String = "ethereum", limit: Int = 50) async throws -> [NFTTrade] {
        var components = URLComponents(string: "http://localhost:8443/api/v1/nft/trades")!
        components.queryItems = [
            URLQueryItem(name: "collection", value: collectionAddress),
            URLQueryItem(name: "chain", value: chain),
            URLQueryItem(name: "limit", value: String(limit))
        ]
        
        let (data, response) = try await session.data(from: components.url!)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            return []
        }
        
        let trades = try JSONDecoder().decode([NFTTrade].self, from: data)
        return trades
    }
}

// MARK: - Error Types

enum NFTError: Error {
    case networkError
    case transactionFailed
    case invalidParameters
    case insufficientBalance
}

// MARK: - Result Types

struct ListingResult: Codable {
    let listingId: String
}

struct OfferResult: Codable {
    let offerId: String
}

struct TransactionResult: Codable {
    let txHash: String
}

// MARK: - Marketplace Enum

enum NFTMarketplace: String, CaseIterable {
    case opensea = "OpenSea"
    case looksRare = "LooksRare"
    case x2y2 = "X2Y2"
    case blur = "Blur"
    
    var displayName: String {
        return self.rawValue
    }
}

File: mobile_apps/flutter_app/lib/services/nft_marketplace_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// NFT Marketplace Service for Flutter App
/// Production-ready NFT marketplace functionality
class NFTMarketplaceService {
  static final NFTMarketplaceService _instance = NFTMarketplaceService._internal();
  factory NFTMarketplaceService() => _instance;
  NFTMarketplaceService._internal();

  final String _baseUrl = 'http://localhost:8443/api/v1/nft';
  
  // Collection Methods
  
  /// Get collection by address
  Future<NFTCollection?> getCollection(String address, {String chain = 'ethereum'}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/collections/$address?chain=$chain'),
      headers: {'Content-Type': 'application/json'},
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return NFTCollection.fromJson(data);
    }
    return null;
  }
  
  /// Search collections
  Future<List<NFTCollection>> searchCollections(String query, {String chain = 'ethereum', int limit = 20}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/collections/search?q=$query&chain=$chain&limit=$limit'),
      headers: {'Content-Type': 'application/json'},
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((e) => NFTCollection.fromJson(e)).toList();
    }
    return [];
  }
  
  // NFT Methods
  
  /// Get NFTs for a collection
  Future<List<NFTItem>> getNFTs(String collectionAddress, {String chain = 'ethereum', int limit = 50, int offset = 0}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/assets?collection=$collectionAddress&chain=$chain&limit=$limit&offset=$offset'),
      headers: {'Content-Type': 'application/json'},
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((e) => NFTItem.fromJson(e)).toList();
    }
    return [];
  }
  
  /// Get NFTs for a wallet
  Future<List<NFTItem>> getUserNFTs(String walletAddress, {String chain = 'ethereum'}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/owners/$walletAddress?chain=$chain'),
      headers: {'Content-Type': 'application/json'},
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((e) => NFTItem.fromJson(e)).toList();
    }
    return [];
  }
  
  // Listing Methods
  
  /// Get listings for a collection
  Future<List<NFTListing>> getListings(String collectionAddress, {String chain = 'ethereum', int limit = 50}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/listings?collection=$collectionAddress&chain=$chain&limit=$limit'),
      headers: {'Content-Type': 'application/json'},
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((e) => NFTListing.fromJson(e)).toList();
    }
    return [];
  }
  
  /// Get user's listings
  Future<List<NFTListing>> getUserListings(String walletAddress, {String chain = 'ethereum'}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/listings/$walletAddress?chain=$chain'),
      headers: {'Content-Type': 'application/json'},
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((e) => NFTListing.fromJson(e)).toList();
    }
    return [];
  }
  
  // Trading Methods
  
  /// Create a listing (sell NFT)
  Future<String> createListing({
    required String walletAddress,
    required String collectionAddress,
    required String tokenId,
    required double price,
    String priceToken = 'ETH',
    String chain = 'ethereum',
  }) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/listings'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'wallet_address': walletAddress,
        'collection_address': collectionAddress,
        'token_id': tokenId,
        'price': price,
        'price_token': priceToken,
        'chain': chain,
      }),
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['listingId'];
    }
    throw Exception('Failed to create listing');
  }
  
  /// Cancel a listing
  Future<bool> cancelListing(String walletAddress, String listingId, {String chain = 'ethereum'}) async {
    final response = await http.delete(
      Uri.parse('$_baseUrl/listings/$listingId'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'wallet_address': walletAddress,
        'chain': chain,
      }),
    );
    
    return response.statusCode == 200;
  }
  
  /// Buy NFT (fulfill listing)
  Future<String> buyNFT(String buyerAddress, String listingId, {String chain = 'ethereum'}) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/purchase'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'buyer_address': buyerAddress,
        'listing_id': listingId,
        'chain': chain,
      }),
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['txHash'];
    }
    throw Exception('Purchase failed');
  }
  
  // Offer Methods
  
  /// Make an offer
  Future<String> makeOffer({
    required String makerAddress,
    required String collectionAddress,
    required String tokenId,
    required double price,
    String priceToken = 'ETH',
    required int expirationTime,
    String chain = 'ethereum',
  }) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/offers'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'maker_address': makerAddress,
        'collection_address': collectionAddress,
        'token_id': tokenId,
        'price': price,
        'price_token': priceToken,
        'expiration_time': expirationTime,
        'chain': chain,
      }),
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['offerId'];
    }
    throw Exception('Failed to create offer');
  }
  
  /// Accept an offer
  Future<String> acceptOffer(String sellerAddress, String offerId, {String chain = 'ethereum'}) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/offers/$offerId/accept'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'seller_address': sellerAddress,
        'chain': chain,
      }),
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['txHash'];
    }
    throw Exception('Failed to accept offer');
  }
  
  // Trade History
  
  /// Get trade history for a collection
  Future<List<NFTTrade>> getCollectionTrades(String collectionAddress, {String chain = 'ethereum', int limit = 50}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/trades?collection=$collectionAddress&chain=$chain&limit=$limit'),
      headers: {'Content-Type': 'application/json'},
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((e) => NFTTrade.fromJson(e)).toList();
    }
    return [];
  }
}

// Models

class NFTCollection {
  final String id;
  final String address;
  final String name;
  final String symbol;
  final String? description;
  final String? imageUrl;
  final String? bannerUrl;
  final double floorPrice;
  final int totalSupply;
  final int owners;
  final double volume24h;
  final String chain;

  NFTCollection({
    required this.id,
    required this.address,
    required this.name,
    required this.symbol,
    this.description,
    this.imageUrl,
    this.bannerUrl,
    required this.floorPrice,
    required this.totalSupply,
    required this.owners,
    required this.volume24h,
    required this.chain,
  });

  factory NFTCollection.fromJson(Map<String, dynamic> json) {
    return NFTCollection(
      id: json['id'] ?? '',
      address: json['address'] ?? '',
      name: json['name'] ?? '',
      symbol: json['symbol'] ?? '',
      description: json['description'],
      imageUrl: json['imageUrl'],
      bannerUrl: json['bannerUrl'],
      floorPrice: (json['floorPrice'] ?? 0).toDouble(),
      totalSupply: json['totalSupply'] ?? 0,
      owners: json['owners'] ?? 0,
      volume24h: (json['volume24h'] ?? 0).toDouble(),
      chain: json['chain'] ?? '',
    );
  }
}

class NFTItem {
  final String id;
  final String tokenId;
  final String collectionAddress;
  final String name;
  final String? description;
  final String? imageUrl;
  final String? animationUrl;
  final List<NFTAttribute>? attributes;
  final String? owner;
  final double? price;
  final String chain;

  NFTItem({
    required this.id,
    required this.tokenId,
    required this.collectionAddress,
    required this.name,
    this.description,
    this.imageUrl,
    this.animationUrl,
    this.attributes,
    this.owner,
    this.price,
    required this.chain,
  });

  factory NFTItem.fromJson(Map<String, dynamic> json) {
    return NFTItem(
      id: json['id'] ?? '',
      tokenId: json['tokenId'] ?? '',
      collectionAddress: json['collectionAddress'] ?? '',
      name: json['name'] ?? '',
      description: json['description'],
      imageUrl: json['imageUrl'],
      animationUrl: json['animationUrl'],
      attributes: json['attributes'] != null 
          ? (json['attributes'] as List).map((e) => NFTAttribute.fromJson(e)).toList()
          : null,
      owner: json['owner'],
      price: json['price']?.toDouble(),
      chain: json['chain'] ?? '',
    );
  }
}

class NFTAttribute {
  final String traitType;
  final String value;

  NFTAttribute({required this.traitType, required this.value});

  factory NFTAttribute.fromJson(Map<String, dynamic> json) {
    return NFTAttribute(
      traitType: json['traitType'] ?? '',
      value: json['value'] ?? '',
    );
  }
}

class NFTListing {
  final String id;
  final String tokenId;
  final String collectionAddress;
  final String seller;
  final double price;
  final String priceToken;
  final int expirationTime;
  final String chain;
  final String? imageUrl;
  final String? name;

  NFTListing({
    required this.id,
    required this.tokenId,
    required this.collectionAddress,
    required this.seller,
    required this.price,
    required this.priceToken,
    required this.expirationTime,
    required this.chain,
    this.imageUrl,
    this.name,
  });

  factory NFTListing.fromJson(Map<String, dynamic> json) {
    return NFTListing(
      id: json['id'] ?? '',
      tokenId: json['tokenId'] ?? '',
      collectionAddress: json['collectionAddress'] ?? '',
      seller: json['seller'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      priceToken: json['priceToken'] ?? 'ETH',
      expirationTime: json['expirationTime'] ?? 0,
      chain: json['chain'] ?? '',
      imageUrl: json['imageUrl'],
      name: json['name'],
    );
  }
}

class NFTTrade {
  final String id;
  final String tokenId;
  final String collectionAddress;
  final String buyer;
  final String seller;
  final double price;
  final int timestamp;
  final String txHash;
  final String chain;

  NFTTrade({
    required this.id,
    required this.tokenId,
    required this.collectionAddress,
    required this.buyer,
    required this.seller,
    required this.price,
    required this.timestamp,
    required this.txHash,
    required this.chain,
  });

  factory NFTTrade.fromJson(Map<String, dynamic> json) {
    return NFTTrade(
      id: json['id'] ?? '',
      tokenId: json['tokenId'] ?? '',
      collectionAddress: json['collectionAddress'] ?? '',
      buyer: json['buyer'] ?? '',
      seller: json['seller'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      timestamp: json['timestamp'] ?? 0,
      txHash: json['txHash'] ?? '',
      chain: json['chain'] ?? '',
    );
  }
}
